rosswhitfield/leeroy
# Avoid initiating CI runs for members who are not in the mantid-developers GitHub team

A validation was added to prevent initiating CI runs for pull requests from unauthorized users who are not part of the GitHub teams defined under `github_teams` in the [config.json](https://github.com/mantidproject/ansible-linode/blob/avoid_ci_runs_for_externals/roles/leeroy/files/config.json#L89-L91) file for Leeroy.

The purpose of this change is to improve the security of the CI environment and reduce the risk of malicious PRs.

When it is necessary to run CI for such pull requests, it is now possible without registering the specific user in the authorized `github_teams`. An already authorized team member can simply add a pull request review comment case-insensitively with "**rerun ci**" to trigger the CI.

**Testing Instructions**
1. Deploy the leeroy service using ansible-linode 
2. Create a test pull request at [mantid-staging](https://github.com/mantidproject/mantid-staging) repository from an account that is not registered at any of the authorized `github_teams`
3. There should be a comment added to the PR by the mantid-builder saying the PR is Unauthorized since CI has not been initialized
4. There need to be a failing check `mantid/unauthorized` with the reason as shown below
<img width="746" height="579" alt="image" src="https://github.com/user-attachments/assets/29a3e388-5819-45c6-a323-ee8b8c4ecb5c" />

5. No CI runs should be initiated against that PR.
6. Put a review comment **`rerun ci`** by an authorized member into the PR as shown below
<img width="579" height="486" alt="image" src="https://github.com/user-attachments/assets/244b98e7-b899-4845-a1da-1c07db07af1d" />


7. A full CI run should now be initiated while marking the status of the previously failed check for `mantid/unauthorized` as successful 
<img width="801" height="678" alt="image" src="https://github.com/user-attachments/assets/d3ca25d7-1246-47f2-8da9-01c3c8caf092" />

8. If the original user pushes further changes, any existing CI builds would be stopped, and an authorised member has to submit **`rerun ci`** review comment each time to initiate the CI. 
9. PRs from authorised members should not see any difference.

The GitHub API used: https://docs.github.com/en/rest/teams/members?apiVersion=2022-11-28#get-team-membership-for-a-user
